lib/src/rules/always_specify_types.dart

```diff
@@ -163,8 +187,51 @@ class _Visitor extends SimpleAstVisitor<void> {
 
   @override
   void visitVariableDeclarationList(VariableDeclarationList list) {
-    if (list.type == null) {
-      rule.reportLintForToken(list.keyword);
+    var keyword = list.keyword;
+    if (list.type == null && keyword != null) {
+      List<String>? types;
+      var parent = list.parent;
+      if (parent is TopLevelVariableDeclaration) {
+        types = _getTypes(parent.variables);
+      } else if (parent is ForPartsWithDeclarations) {
+        types = _getTypes(parent.variables);
+      } else if (parent is FieldDeclaration) {
+        types = _getTypes(parent.fields);
+      } else if (parent is VariableDeclarationStatement) {
+        types = _getTypes(parent.variables);
+      }
+
+      if (types == null) return;
+
+      String? multipleTypesString;
```

Comment:
If there are multiple types then it's highly likely that none of them are correct for all the cases. What we really need to do in that case is to (a) capture the `DartType` object for each of the types rather than the names of the types and (b) use the `TypeSystem` to find the least upper bound of those types. We can then suggest the one correct type.

To do even better, if there's more than one type we might want a different correction message (which implies a different instance of `LintCode`) that read more like "Try specifying the type '{0}', or split the declaration into multiple declarations in order to use more specific types.".

Reply:
Thanks for feedback, I didn't think about this.

However, I don't think we can simply use the upper bound of those types, because we must first ensure no sub-class methods are being invoked, and I am not sure if this is possible.

Suppose we have the following case

```dart
class A {
  void a() {}
}
class B extends A {
  void b() {}
}
void main() {
  var a = A(), b = B();
  a.a();
  b.b();
}
```

We can't simply say it can be changed to

```dart
A a = A(), b = B();
```

Because `b.b()` is not part of `A`.

Is there a way to know if all method invocations of `a` and `b` are relevant to only `class A` and not `B`?

Reply:
You're right, the least upper bound isn't sufficient in this case. The only guaranteed solution is to split the declarations. So, let's not provide any type in the message and shorten the correction message to only include the one valid option (something like "Try splitting the declaration into multiple declarations.").